src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -685,27 +916,164 @@ Value emitScalarOpFor<ONNXRoundOp>(ConversionPatternRewriter &rewriter,
   // }
   // return y;
   // ```
-  Value one = createMath.constant(elementType, 1.0);
-  Value two = createMath.constant(elementType, 2.0);
-  Value half = createMath.constant(elementType, 0.5);
-  Value y = createMath.floor(x);
-  Value r = createMath.sub(x, y);
+  Value one = create.math.constant(elementType, 1.0);
+  Value two = create.math.constant(elementType, 2.0);
+  Value half = create.math.constant(elementType, 0.5);
+  Value y = create.math.floor(x);
+  Value r = create.math.sub(x, y);
   // r > 0.5
-  Value rGreaterThanHalf = createMath.sgt(r, half);
-  Value y1 = createMath.select(rGreaterThanHalf, createMath.add(y, one), y);
+  Value rGreaterThanHalf = create.math.sgt(r, half);
+  Value y1 = create.math.select(rGreaterThanHalf, create.math.add(y, one), y);
   // r == 0.5: round to nearest even.
-  Value y2 = createMath.mul(half, y);
-  y2 = createMath.floor(y2);
-  y2 = createMath.mul(y2, two);
-  Value rr = createMath.sub(y, y2);
-  Value rrEqualOne = createMath.eq(rr, one);
-  y2 = createMath.select(rrEqualOne, createMath.add(y, one), y);
-
-  Value rEqualHalf = createMath.eq(r, half);
-  return createMath.select(rEqualHalf, y2, y1);
+  Value y2 = create.math.mul(half, y);
+  y2 = create.math.floor(y2);
+  y2 = create.math.mul(y2, two);
+  Value rr = create.math.sub(y, y2);
+  Value rrEqualOne = create.math.eq(rr, one);
+  y2 = create.math.select(rrEqualOne, create.math.add(y, one), y);
+
+  Value rEqualHalf = create.math.eq(r, half);
+  return create.math.select(rEqualHalf, y2, y1);
+}
+
+//===----------------------------------------------------------------------===//
+// SIMD code gen for kernels where data can be fully flattened.
+//===----------------------------------------------------------------------===//
+
+using MDBuilder = MultiDialectBuilder<IndexExprBuilderForKrnl, KrnlBuilder,
+    MemRefBuilder, VectorBuilder>;
+
+//
+template <typename ElementwiseUnaryOp>
+static LogicalResult getUnaryBinarySimdCodeFullyFlattened(
+    ConversionPatternRewriter &rewriter, MDBuilder &create,
+    ONNXOpShapeHelper *shapeHelper, Operation *op, MemRefType outputMemRefType,
+    ArrayRef<Value> operands, int64_t alignment, int64_t simdUnroll) {
+  Type outputElementType = outputMemRefType.getElementType();
+
+  // generate SIMD code of VL elements per vector.
+  IndexExprScope allocScope(create.vec, shapeHelper->getScope());
+  int64_t VL =
+      create.vec.getMachineVectorLength(outputElementType) * simdUnroll;
+  // Alloc memory with padding for SIMD.
+  Value alloc = create.mem.alignedAllocWithSimdPadding(
+      outputMemRefType, shapeHelper->getOutputDims(), simdUnroll, alignment);
```

Comment:
Now I understand that you create a padded buffer first then view it as the original shape. That's fine. I though you created a buffer of the original shape and viewed it as a padded buffer when looking at your slide at our meeting.